[all]: Migrate eslint to version 9 and refactor accordingly

MenKNas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

- [ ] This PR follows the [Contribution Guide](https://github.com/Sitecore/content-sdk/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated
- [ ] Upgrade guide entry added

## Description / Motivation

This PR migrates ESLint from legacy configuration to modern flat config format, updating to the latest stable versions.

Core Changes:

- ESLint 9 Migration: Upgraded ESLint `8.57.0 → 9.0.0` and migrated from deprecated `.eslintrc` files to modern `eslint.config.mjs` flat config

- Package Dependencies: Updated all ESLint packages to the latest stable versions and added missing plugins (eslint-plugin-import, eslint-plugin-react-hooks)

- Configuration Consolidation: Replaced all package-level `.eslintrc` files with centralized flat config using file patterns for package-specific TypeScript and React rules


## Testing Details
- [ ] Unit Test Added
- [x] Manual Test/Other => Tested the correct linting of all packages with `yarn lint-packages` command after all the changes. Also tried some scenarios that triggered linting errors and were correctly caught by the linter.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

MenKNas committed:
[all]: Migrate eslint to version 9 and refactor accordingly

MenKNas committed:
[all]: Version fixes and small updates

MenKNas committed:
[all]: fix linting issues

MenKNas committed:
remove yarn lock

MenKNas committed:
update yarn lock

MenKNas committed:
update yarn lock

MenKNas committed:
[all] Fix dependency issues

MenKNas committed:
Merge branch 'dev' into feature/JSS-4780-migrate-to-flat-eslint

MenKNas committed:
[all] Fix dependency issues

MenKNas committed:
[all] Adjust linting rules in various files with issues

MenKNas committed:
[all] Adjust linting rules in various files with issues

MenKNas committed:
[all] Adjust linting rules in various files with issues

MenKNas committed:
[all] Adjust linting rules in various files with issues

MenKNas committed:
[all] Adjust linting rules in various files with issues

MenKNas committed:
[all]: Fix failing tests

MenKNas committed:
[all]: Fix scaffolded app issues

MenKNas committed:
Revert "[all]: Fix scaffolded app issues"

This reverts commit daa3624d54502582c200dcccd81485501aeac26e.

MenKNas committed:
[all]: Add missing plugings for eslint config

MenKNas committed:
[all]: Fix other linting issues

MenKNas committed:
[all]: Fix other linting issues

MenKNas committed:
[all]: Fix other linting issues

mg-aceik commented on packages/create-content-sdk-app/package.json:
```diff
@@ -38,6 +38,7 @@
     "minimist": "^1.2.8"
   },
   "devDependencies": {
+    "@stylistic/eslint-plugin-ts": "^2.10.0",
```
If you're going to this effort, I don't understand why not use the latest v5 @stylistic/eslint-plugin? You're bringing in a whole lot of old dependencies here. 

Also I liked how un-opinionated it was about eslint config before, and now this is going to make it tricker to understand and remove your opinions when using this template.

(I meant to comment this on the `packages\create-content-sdk-app\src\templates\nextjs\package.json` but the same issue is there too).

MenKNas replied:
@mg-aceik Thanks for the feedback. This is still a work in progress and not finalized yet.

MenKNas committed:
[all]: Update packages

MenKNas committed:
[all]: Update packages

MenKNas committed:
[all]: Update packages

MenKNas committed:
[all]: Update packages

MenKNas committed:
[all]: Update yarn.lock

MenKNas committed:
[all]: Upgrade react related linting packages to fix issues

illiakovalenko committed:
[Chore] Test non npm-token publish

Commit:
Update API docs

Commit:
v1.1.0-canary.19 [skip ci]

MenKNas committed:
Merge branch 'dev' into feature/JSS-4780-migrate-to-flat-eslint

illiakovalenko committed:
[Chore] Test non token publish

MenKNas committed:
[all]: Fix yarn.lock issues

MenKNas committed:
Merge branch 'dev' into feature/JSS-4780-migrate-to-flat-eslint

Commit:
Update API docs

MenKNas committed:
[all]: Resolve conflicts

MenKNas committed:
[all]: Resolve conflicts

MenKNas committed:
[all]: Resolve conflicts

MenKNas committed:
[all]: Resolve conflicts

MenKNas committed:
[all]: Fix yarn.lock file

MenKNas committed:
[all]: Restore publish.yml file

MenKNas commented:
@Sitecore/jss-dev-reviewers 

Hello team,

This PR is now ready for a first round of review. Some things to notice and discuss further are the following:

- The approach I followed after migrating to `ESLint 9` is to have a centralized file `eslint.config.mjs` that follows the latest _**flat**_ syntax at the root of our project and is responsible for all our linting rules and logic. We can use this file to extend our linting rules for other packages without the obligation to maintain a dedicated file for each package. I find this approach more flexible and maintainable moving forward, but I am open to further discussion or other suggestions.

- In most cases I tried to keep our previous linting rules but also added some suggested/modern ones that will make our code easier to read. Of course this is personal preference so I would like to hear your opinion as well. We can adjust the rules according to our needs.

- Something weird happened after I resolved conflicts, and I can see all `.md` files added as changed because previously a final line for each file was removed. All files should be exactly the same as `dev` now so I think they can be ignored but it wouldn't hurt to have an extra quick look for those in case something was missed.

art-alexeyenko commented on .github/workflows/publish.yml:
```diff
@@ -65,4 +65,4 @@ jobs:
           npx lerna publish from-git --exact --canary --pre-dist-tag canary --preid canary --no-verify-access --yes
         env:
-          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
```
probably an error here, removing empty line

MenKNas replied:
It was just removed before (not sure why) and re-added so it should work fine (looks exactly the same as `dev`)

art-alexeyenko commented on tsconfig.json:
```diff
@@ -17,5 +17,8 @@
-  }
+  },
+  "exclude": [
+    "packages/create-content-sdk-app/src/templates/**/*"
```
What does this change do?

MenKNas replied:
This change helps us exclude `src/templates` and keeps the TS compiler and editor from type-checking our boilerplate template files which would produce spurious errors. ESLint already ignores them, but TS needs its own exclude.

art-alexeyenko commented on eslint.config.mjs:
```diff
@@ -0,0 +1,155 @@
+const compat = new FlatCompat({
+  baseDirectory: __dirname,
+  recommendedConfig: js.configs.recommended,
+  allConfig: js.configs.all,
```
I'd like to confirm something. In original eslintrc we have the list of recommended configs being extended:
```
"extends": [
    "eslint:recommended",
    "plugin:jsdoc/recommended",
    "prettier",
    "prettier/@typescript-eslint"
  ],
```
How are these being replaced in flat config?

MenKNas replied:
The syntax has been updated and now accomodates all the previous configs plus new expanded rules

mg-aceik commented on packages/create-content-sdk-app/src/templates/nextjs/package.json:
```diff
@@ -42,16 +42,19 @@
+    "eslint-plugin-react-hooks": "5.2.0",
     "npm-run-all2": "~8.0.1",
-    "prettier": "^3.5.3",
+    "prettier": "^2.8.8",
```
Why the downgrade?

mg-aceik commented on packages/create-content-sdk-app/src/templates/nextjs/.eslintrc:
Maybe I'm missing something but now the template won't include any eslint config?

MenKNas replied:
A new `eslint.config` file has also been introduced for templates, following the same latest flat config syntax.

mg-aceik commented on packages/create-content-sdk-app/src/templates/nextjs/package.json:
```diff
@@ -42,16 +42,19 @@
+    "@typescript-eslint/parser": "8.39.0",
     "cross-env": "~7.0.3",
-    "eslint": "^8.56.0",
-    "eslint-config-next": "^13.1.5",
```
Have you considered replicating the same eslint packages and config you get from running `npx create-next-app@latest`?

MenKNas replied:
These have been updated.

MenKNas committed:
[all]: Package and configs updates

MenKNas committed:
[all]: Resolve conflicts

MenKNas committed:
[all]: Fix broken package.json files

MenKNas committed:
[all]: Fix linting issue

MenKNas committed:
[all]: Fix issues with scaffolded samples

MenKNas committed:
[all]: Fix issues with scaffolded samples

MenKNas committed:
Merge branch 'dev' into feature/JSS-4780-migrate-to-flat-eslint

MenKNas committed:
[all]: Fix yarn lock issues

addy-pathania commented on eslint.config.mjs:
```diff
@@ -0,0 +1,175 @@
+import { defineConfig, globalIgnores } from 'eslint/config';
+// import { FlatCompat } from '@eslint/eslintrc';
```
is this comment intentional?

addy-pathania commented on eslint.config.mjs:
```diff
@@ -0,0 +1,175 @@
+        },
+      ],
+      'react-hooks/rules-of-hooks': 'error',
+      'react-hooks/exhaustive-deps': 'warn',
```
This looks repeated. Above you're setting this to `off`.

addy-pathania commented on eslint.config.mjs:
```diff
@@ -0,0 +1,175 @@
+      'spaced-comment': 'error',
+      curly: ['error', 'multi-line'],
+      'eol-last': ['error', 'always'],
+      'linebreak-style': ['error', 'windows'],
```
should we remove this to avoid cross-OS issues?

addy-pathania reviewed (commented):
Great work. See some comments.

MenKNas committed:
[all] Address PR comments + minor enhancements

MenKNas commented:
> Great work. See some comments.

Thanks @addy-pathania, I addressed the comments and also expanded our `eslint.config` file slightly. Feel free to have a final look.

MenKNas committed:
Merge branch 'dev' into feature/JSS-4780-migrate-to-flat-eslint

MenKNas committed:
[all] Fix yarn.lock issue

addy-pathania commented on eslint.config.mjs:
```diff
@@ -0,0 +1,172 @@
+      radix: 'error',
+      'default-case': 'error',
+      eqeqeq: 'error',
+      'jsx-quotes': ['error', 'prefer-double'],
```
In jsx we want double quotes?

MenKNas replied:
Yes, it's preferable because of attributes like `<Button label="OK" />` or similar. Prettier already handles this, but it doesn't hurt to enforce it extra.

addy-pathania commented on eslint.config.mjs:
```diff
@@ -0,0 +1,172 @@
+      '@typescript-eslint/no-require-imports': 'off',
+      '@typescript-eslint/no-explicit-any': 'off',
+      '@typescript-eslint/no-unused-expressions': 'off',
+      '@typescript-eslint/no-unused-vars': 'off',
```
this is again listed twice or there is some reason?

MenKNas replied:
In fact there aren’t duplicate keys in the same rules block. What looks duplicated are the core JS rules vs the TS plugin rules.

In the base config we use core rules (e.g. no-unused-vars, no-unused-expressions) for JS.

In the TS override we explicitly turn off the TS equivalents (@typescript-eslint/no-*) to avoid overlap and keep the current behavior unchanged.

addy-pathania commented on eslint.config.mjs:
```diff
@@ -0,0 +1,172 @@
+          custom: { regex: '^I[A-Z]', match: false },
+        },
+      ],
+      'no-undef': 'off',
```
Since TypeScript has its own type checking, this rule might not be necessary. But its upto you to remove.

MenKNas replied:
I would keep this because core no-undef misfires in TypeScript, and the TS compiler already catches undefined identifiers. Leaving `'no-undef': 'off'` in your TS override prevents noisy false positives.

addy-pathania replied:
make sense

addy-pathania reviewed (approved):
👍🏼
